GitCommands/Git/IndexLockManager.cs

```diff
@@ -0,0 +1,135 @@
+﻿using System;
+using System.IO;
+using System.IO.Abstractions;
+using System.Linq;
+using GitUIPluginInterfaces;
+
+namespace GitCommands.Git
+{
+    public interface IIndexLockManager
+    {
+        /// <summary>
+        /// Determines whether the given repository has index.lock file.
+        /// </summary>
+        /// <param name="repositoryPath"></param>
+        /// <returns><see langword="true"/> is index is locked; otherwise <see langword="false"/>.</returns>
+        bool IsIndexLocked(string repositoryPath);
+
+        /// <summary>
+        /// Delete index.lock in the current working folder.
+        /// </summary>
+        /// <param name="includeSubmodules">
+        /// If <see langword="true"/> all submodules will be scanned for index.lock files and have them delete, if found.
+        /// </param>
+        /// <exception cref="FileDeleteException">Unable to delete specific index.lock.</exception>
+        void UnlockIndex(bool includeSubmodules);
```

Comment:
I find it little inconsistent. IsIndexLocked takes repositoryPath as an argument but UnlockIndex does not. I think that either both should take repoPath as an argument or none of them.

Reply:
A fair call, I overlooked that.
Thinking of it, it probably makes sense to modify the signature to `void UnlockIndex(string repositoryPath, bool includeSubmodules)`

It annoys me that I can't reset/update submodules if they are locked - I think this change should allow to address that (I will have to test it).

Reply:
Giving it a second thought I realized that it should not take a path since it depends on GitModule.
If the given path is for a different repo then the unit tests will not make sense. If the path is always the working dir of the provided GitModule then there is no sense to pass it to the UnlockIndex method. So IndexLockManager either should depend on GitModule or GitModule should be passed to the all IndexLockManager methods.

Reply:
The manager depends on `IGitModule` (injected in).
The methods could take optional `repositoryPath` argument and fall back to the `IGitModule.WorkingDir` if none supplied.
Anyway I will review the API again (when I have a chance) before merging it in

Reply:
How does the following look to you?

```
    public interface IIndexLockManager
    {
        /// <summary>
        /// Determines whether the given repository has index.lock file.
        /// </summary>
        /// <param name="repositoryPath">The repository path. Defaults to the working directory if none supplied.</param>
        /// <returns><see langword="true"/> is index is locked; otherwise <see langword="false"/>.</returns>
        bool IsIndexLocked(string repositoryPath = null);

        /// <summary>
        /// Delete index.lock in the current working folder.
        /// </summary>
        /// <param name="repositoryPath">The repository path. Defaults to the working directory if none supplied.</param>
        /// <param name="includeSubmodules">
        ///     If <see langword="true"/> all submodules will be scanned for index.lock files and have them delete, if found.
        /// </param>
        /// <exception cref="FileDeleteException">Unable to delete specific index.lock.</exception>
        void UnlockIndex(string repositoryPath = null, bool includeSubmodules = true);
    }

```

Reply:
I don't know how you are going to rewrite tests using IGitModule injected in to the manager:
https://github.com/gitextensions/gitextensions/pull/4065/files#diff-1575c34edaf5ecacc8a250a1c8a2d159R80